### PR TITLE
[Inductor Intel GPU backend Upstream] Step 1/3

### PR DESCRIPTION
**Pull request by etaf**
This is the step 1 to add Intel GPU backend as an alternative inductor backend.
Feature request: https://github.com/pytorch/pytorch/issues/114856.

To allow the Intel GPU backend to utilize existing code as much as possible, this PR generialized the device-biase code in triton based code generation. The PR can be summarized as:
- We created a DeviceOverrides interface class for device specific code generation in `WrapperCodegen` and `TritonScheduling`.
With that, a new backend can easily re-use`WrapperCodegen` and `TritonScheduling` by inherit them and set their own DeivceOverrides for their device flexibility.
- We generilized Inductor some particular runtime functions per device during generation process using  [general device interface
](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/device_interface.py#L34)

### CODE REVIEW

File: torch/_inductor/codegen/common.py [hunk 1]
```diff
@@ -450,6 +450,39 @@ def remainder(a, b):
     def load_seed(name, offset):
         return ops.load(name, sympy.Integer(offset))
 
+class DeviceOverrides():
```
Comment (EikanWang) [inline torch/_inductor/codegen/common.py:453]: How about changing it to `WrapperOpsOverrides`

File: torch/_inductor/codegen/common.py [hunk 2]
```diff
@@ -450,6 +450,39 @@ def remainder(a, b):
...
+
+    @classmethod
+    def py_import_get_raw_stream_as(self, name):
```
Comment (EikanWang) [inline torch/_inductor/codegen/common.py:456]: Please remove `py_` prefix.

File: torch/_inductor/codegen/common.py [hunk 3]
```diff
@@ -450,6 +450,39 @@ def remainder(a, b):
...
+    @classmethod
+    def py_DeviceGuard(self, device_idx):
+        raise NotImplementedError()
+
+    @classmethod
+    def cpp_defAOTStreamGuard(self, name, stream, device_idx):
+        raise NotImplementedError()
+
+    @classmethod
+    def cpp_defStreamGuard(self, name, stream):
```
Comment (EikanWang) [inline torch/_inductor/codegen/common.py:476]: CPP is not a current priority. Let's remove it first.

File: torch/_inductor/triton_heuristics.py [hunk 4]
```diff
@@ -50,11 +50,6 @@
     KernelInterface = object
     OutOfResources = object
 
-if has_triton():
```
Comment (EikanWang) [inline torch/_inductor/triton_heuristics.py:53]: Why do we need to remove the check?

File: torch/_inductor/codegen/triton.py [hunk 5]
```diff
@@ -818,6 +819,7 @@ def __init__(
         pid_cache=None,
         reduction_hint=ReductionHint.DEFAULT,
         min_elem_per_thread=0,
+        deviceOverrides=CUDADeviceOverrides
```
Comment (EikanWang) [inline torch/_inductor/codegen/triton.py:822]: We added multiple `DeviceOverrides`. Let's add it to the `Virtualized` and set `DeviceOverrides` according to the graph information.

File: torch/_inductor/triton_heuristics.py [hunk 6]
```diff
@@ -152,6 +147,8 @@ def __init__(
         self.mutated_arg_names = mutated_arg_names
         self.configs = configs
         self.heuristic_type = heuristic_type
+        self.device_type = triton_meta['device_type']
+        self.device_interface = get_interface_for_device(self.device_type)
```
Comment (EikanWang) [inline torch/_inductor/triton_heuristics.py:151]: Please ensure this PR focuses on `WrapperCodeGen` generalization but do not need to touch non-wrapper-code-gen.

**Review by EikanWang — CHANGES_REQUESTED** [review]

**Comment by EikanWang** [comment]
Please format your code.

File: torch/_inductor/codegen/common.py (hunk 3 above)
Comment (etaf) [reply torch/_inductor/codegen/common.py:476]: CUDA backend need this in EnterCudaDeviceContextManagerLine, [line](https://github.com/etaf/pytorch-inductor-xpu/blob/b8ce05456cd6e332746165e469ab1e9048c3daf5/torch/_inductor/codegen/wrapper.py#L198C34-L198C34) , do you mean we should not remove all cuda-bias code in EnterCudaDeviceContextManagerLine this time?

Comment (etaf) [reply torch/_inductor/codegen/common.py:476]: > CPP is not a current priority. Let's remove it first.

CUDA backend need this in EnterCudaDeviceContextManagerLine, [line](https://github.com/etaf/pytorch-inductor-xpu/blob/b8ce05456cd6e332746165e469ab1e9048c3daf5/torch/_inductor/codegen/wrapper.py#L198C34-L198C34) , do you mean we should not remove all cuda-bias code in EnterCudaDeviceContextManagerLine this time?

File: torch/_inductor/codegen/common.py (hunk 1 above)
Comment (etaf) [reply torch/_inductor/codegen/common.py:453]: > How about changing it to `WrapperOpsOverrides`

The interface also used by TritonScheduling, is it OK to change to  `WrapperOpsOverrides`?

File: torch/_inductor/triton_heuristics.py (hunk 4 above)
Comment (etaf) [reply torch/_inductor/triton_heuristics.py:53]: > Why do we need to remove the check?

Here they use `get_cuda_stream` from [triton](https://github.com/openai/triton/blob/v2.1.0/python/triton/runtime/jit.py#L20C31-L20C31)
Talked with guangye, the get_cuda_stream in triton is a workaround for compatibilitythe old version that `_cuda_getCurrentRawStream` has not implemented. 
Currently we can directly use `_cuda_getCurrentRawStream`.

File: torch/_inductor/codegen/common.py (hunk 1 above)
Comment (EikanWang) [reply torch/_inductor/codegen/common.py:453]: Make sense. Then how about changing it to `DeviceOpsOverrides`?

File: torch/_inductor/codegen/common.py (hunk 3 above)
Comment (EikanWang) [reply torch/_inductor/codegen/common.py:476]: How about adding `namespace` as a parameter? Its value can be either `aten` or `torch` and the default is `torch`.

**Pull request closed without merging**